### DISCUSSION TITLE
Return appropriate exception when using a tx_hash which is not using uniswap router

### DISCUSSION THREAD

**Discussion by 1mBC** [Elnaril/uniswap-universal-router-decoder#68] [Ideas]
I use this library to inspect many transactions and some of them are not made on uniswap. I would love to have a specific error response when it is the case.

**Comment by Elnaril** [comment]
Hi @1mBC and thank you for your interest in this lib! :)

Regarding your suggestion, could you please share an example of such transactions?

**Reply by 1mBC** [reply]
yes, this one for example : 

0xf6f6ca0bb63a0f55b686a6982a8ae49136c95e58b9991e36be2c516ad056a528

**Reply by Elnaril** [reply]
Ok, thanks!
In this case, why not just check whether the transaction recipient is the Uniswap UR address or not?

**Reply by 1mBC** [reply]
it is  for sure what i am doing for now :)
but it forces me to do a w3.eth.get_transaction (either before or after a fail) which I would suggest I shouldn't have to do, for performance and elegance reasons :)
I haven't looked but I guess you already do this get_transaction and should check and raise the exception instead I get a "Could not find any function with matching selector"

**Reply by Elnaril** [reply]
>  it forces me to do a w3.eth.get_transaction (either before or after a fail) which I would suggest I shouldn't have to do, for performance and elegance reasons :)

Ok, I think I see your point. 👍

The `ValueError` "Could not find any function with matching selector" is raised by the web3.py lib and is really just that: the ABI does not contain a function corresponding to the selector in the input data. Although it is likely to mean the input data is not for the UR, it would be misleading to change the message to something of higher perspective (like "Not the Uniswap Universal Router"), just based on this exception. Beside, I prefer to keep exceptions that are well known from web3.py users.

The other way would be to check the smart contract address. It would mean that all UR addresses are listed in the lib _for all blockchains_ ! That would be hardly maintainable ...

And checking if the transaction is sent to the UR without calling twice the `w3.eth.get_transaction()` method (one in the lib and one in your code) can be achieved by calling it once in your code and then by using `codec.decode.function_input()` i/o `codec.decode.transaction()`.

Hope it makes sense? :)

**Reply by 1mBC** [reply]
maybe I am bit a noob on the matter, so forgive me if I'm saying nonsense, but I use your lib with transaction hash, so I don't have the tx input before calling decode.transaction() so I could use decode.function_input()
if that makes sense

And I don't think it would behardly maintainable to just check if the "to" is the uniswap universal router.

But I get your point on prefering to keep the web3 strandards
